swarm/network/stream/snapshot_sync_test.go

```diff
@@ -0,0 +1,734 @@
+package stream
+
+import (
+	"context"
+	crand "crypto/rand"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/simulations"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/pot"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/network"
+	streamTesting "github.com/ethereum/go-ethereum/swarm/network/stream/testing"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+const testMinProxBinSize = 2
+const MAX_TIMEOUT = 600
+
+var (
+	pof = pot.DefaultPof(256)
+
+	conf      *synctestConfig
+	startTime time.Time
+	ids       []discover.NodeID
+	datadirs  map[discover.NodeID]string
+	ppmap     map[discover.NodeID]*network.PeerPot
+
+	live    bool
+	history bool
+
+	longrunning = flag.Bool("longrunning", false, "do run long-running tests")
+)
+
+type synctestConfig struct {
+	addrs            [][]byte
+	chunks           []storage.Key
+	retrievalMap     map[string]map[string]time.Duration
+	idToChunksMap    map[discover.NodeID][]int
+	chunksToNodesMap map[string][]int
+	idToAddrMap      map[discover.NodeID][]byte
+	addrToIdMap      map[string]discover.NodeID
+}
+
+func init() {
+	rand.Seed(time.Now().Unix())
+}
+
+//common_test needs to initialize the test in a init() func
+//in order for adapters to register the NewStreamerService;
+//this service is dependent on some global variables
+//we thus need to initialize first as init() as well.
+func initSyncTest() {
+	//assign the toAddr func so NewStreamerService can build the addr
+	toAddr = func(id discover.NodeID) *network.BzzAddr {
+		addr := network.NewAddrFromNodeID(id)
+		// addr.OAddr[0] = byte(0)
+		return addr
+	}
+
+	createStoreFunc = createTestLocalStorageForId
+	//local stores
+	stores = make(map[discover.NodeID]storage.ChunkStore)
+	//data directories for each node and store
+	datadirs = make(map[discover.NodeID]string)
+	//deliveries for each node
+	deliveries = make(map[discover.NodeID]*Delivery)
+	//registries, map of discover.NodeID to its streamer
+	registries = make(map[discover.NodeID]*TestRegistry)
+
+	//channel to wait for peers connected
+	//not needed for this test but required from common_test for NewStreamService
+	waitPeerErrC = make(chan error)
+	//also not needed for this test but required for NewStreamService
+	peerCount = func(id discover.NodeID) int {
+		if ids[0] == id || ids[len(ids)-1] == id {
+			return 1
+		}
+		return 2
+	}
+}
+
+//This file executes a number of tests with the syntax
+//TestSyncing_x_y
+//x is the number of chunks which will be uploaded
+//y is the number of nodes for the test
+func TestSyncing_4_32(t *testing.T)  { testSyncing(t, 4, 64) }
+func TestSyncing_32_16(t *testing.T) { testSyncing(t, 128, 16) }
+
+func TestLongRunningSyncing(t *testing.T) {
```

Comment:
how about some for loop to have some consistency.
OTOH, I am not sure we need all these